text/3371-cargo-target-dir-templates.md

```diff
@@ -0,0 +1,423 @@
+
+A `cargo build` in `perso-1` produces new artifacts in `/cargo-cache/<manifest-path-hash>/debug/...`.
+
+A `cargo clean` only removed the `/cargo-cache/<manifest-path-hash>/` subdirectory, not all the artifacts for all other projects that are also in the cache.
+
+In this situation, it's much less likely for Cargo to produce invalid state without a `build.rs` deliberately writing outside its target directory.
+
+Two projects can be built in parallel without troubles.
+
+#### Absolute and relative paths
+
+`CARGO_TARGET_DIR` can be either a relative or absolute path, which makes sense since it's mostly intended for a single project, which can then work from its own position to configure the target directory, and that stays the case with templates.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+-->
+
+## Templating values of `CARGO_TARGET_DIR`
+
+Templating does not interfere with the resolution order of `CARGO_TARGET_DIR`. From less to most specific:
+
+- Through the `config.toml`:
+
+  ```toml
+  [build]
+  target-dir = "/absolute/path/to/cache/{manifest-path-hash}"
+  ```
+
+- Through the environment variable: `CARGO_TARGET_DIR="/absolute/path/to/cache/{manifest-path-hash}" cargo build`
+- Through the command line flag: `cargo build --target-dir "/absolute/path/to/cache/{manifest-path-hash}"`
+
+## Naming
+
+In the example in the previous section, using a templated `CARGO_TARGET_DIR` with `cargo build` produces named subdirectories. The name of those is computed from the full and canonicalized path to the manifest for the workspace, so building `work-project/crate-1` will still use the directory for the whole workspace during a `cargo build` call.
+
+This naming scheme is **not considered stable**: the method will probably not change often but `cargo` offers no guarantee and may change it in any release. Tools that needs to interact with `cargo`'s target directory should not rely on its value for more than a single invocation of them: they should instead query `cargo metadata` for the actual value each time they are invoked.
+
+The path used for the naming of the final target directory is the one found *after* symlink resolution: `bazel` does it too and I have not found any complaints about this and it has the distinct advantage of allowing to make a symlink to a project somewhere else on the machine (for example to organise work projects) and avoid duplicating the build directory and all its data (which can be quite heavy).
+
```

Comment:
I feel like the references to `naming` in here should be changed to refer to "`{manifest-path-hash}`"

```suggestion
## Definition of `{manifest-path-hash}`

In the example in the previous section, `{manifest-path-hash}` was replaced with a relative path. This relative path is computed from the full and canonicalized path to the manifest for the workspace `Cargo.toml`.
By being canonicalized, including resolving of symlinks, symlinked projects will share the same target directory.  This is following the prior art from `bazel` and I have not found any complaints about this.

The hashing and turning the hash into nested directories is **not considered stable**: the method will probably not change often but `cargo` offers no guarantee and may change it in any release. Tools that needs to interact with `cargo`'s target directory should not rely on its value for more than a single invocation of them: they should instead query `cargo metadata` for the actual value each time they are invoked.
```

Reply:
Done in 81bac72702e64fe03a0cfaf3050e71d7383922e5